librosa/core/audio.py

```diff
@@ -357,14 +357,20 @@ def get_duration(y=None, sr=22050, S=None, n_fft=2048, hop_length=512,
     -------
     d : float >= 0
         Duration (in seconds) of the input time series or spectrogram.
+
+    Raises
+    ------
+    ParameterError
+        if none of `y`, `S`, or `filename` are provided.
     """
 
     if filename is not None:
         with audioread.audio_open(filename) as fdesc:
             return fdesc.duration
 
     if y is None:
-        assert S is not None
+        if S is None:
+            raise ParameterError('At least one of (y, sr), S, or filename must be provided')
```

Comment:
I'm a bit circumspect with this `at least one of X, Y, Z` design
where `X` is `(y,sr)`, `Y` is `S` and `Z` is `filename`
why not make these three options mutually exclusive?
at the moment `Z` takes precedence over `Y`, which takes precedence over `X`
this is undocumented

Reply:
Making them mutually exclusive would be more complicated to implement.  But the precedence should definitely be documented here.

Reply:
making them mutually exclusive has a nice "single dispatch" flavor

i really don't imagine a use case for a callsite of the form
`get_duration(y=my_y, sr=my_sr, S=my_S, filename=my_filename)`
in which, some, but not all, of `my_y`, `my_S` and `my_filename`, would be `None` at runtime in user code.

but this is not a hill i'm willing to die upon. i'm fine with anything as long as it's properly documented

Reply:
> i really don't imagine a use case for a callsite of the form
> `get_duration(y=my_y, sr=my_sr, S=my_S, filename=my_filename)`
> in which, some, but not all, of `my_y`, `my_S` and `my_filename`, would be `None` at runtime in user code.

Multiple dispatch over `(y, sr)` and `S` is pervasive in librosa, for better or worse.  A user wouldn't necessarily call it as you've described, but for interdepdent functions (say, `mfcc` and `melspectrogram`), it's handy to have a nice interface for directly passing through the different forms of input.

`get_duration` started life in that form (for the sake of stylistic consistency), and the `filename` option was added later.

Reply:
Already done! :tada: